pkg/controller/elasticsearch/driver/upscale.go

```diff
@@ -66,35 +72,112 @@ func HandleUpscaleAndSpecChanges(
+	// which should happen 1 at a time as we adjust the replicas early.
+	if err = maybeUpscaleMasterResources(ctx, masterResources); err != nil {
+		return results, fmt.Errorf("while scaling up master resources: %w", err)
+	}
+
+	// First, reconcile all non-master resources
+	results, err = reconcileResources(ctx, actualStatefulSets, nonMasterResources)
+	if err != nil {
+		return results, fmt.Errorf("while reconciling non-master resources: %w", err)
+	}
+	results.ActualStatefulSets = actualStatefulSets
+
+	if results.Requeue {
+		return results, nil
+	}
+
+	targetVersion, err := version.Parse(ctx.es.Spec.Version)
+	if err != nil {
+		return results, fmt.Errorf("while parsing Elasticsearch upgrade target version: %w", err)
+	}
+
+	// Check if all non-master StatefulSets have completed their upgrades before proceeding with master StatefulSets
+	pendingNonMasterSTS, err := findPendingNonMasterStatefulSetUpgrades(
+		ctx.k8sClient,
+		actualStatefulSets,
+		expectedResources.StatefulSets(),
+		targetVersion,
+		ctx.expectations,
+	)
+	if err != nil {
+		return results, fmt.Errorf("while checking non-master upgrade status: %w", err)
+	}
+
+	ctx.upscaleReporter.RecordPendingNonMasterSTSUpgrades(pendingNonMasterSTS)
+
+	if len(pendingNonMasterSTS) > 0 {
+		// Non-master StatefulSets are still upgrading, skipping master StatefulSets temporarily.
+		// This will cause a requeue in the caller, and master StatefulSets will attempt to be processed in the next reconciliation
+		return results, nil
+	}
+
+	// All non-master StatefulSets are upgraded, now process master StatefulSets
+	results, err = reconcileResources(ctx, actualStatefulSets, masterResources)
+	if err != nil {
+		return results, fmt.Errorf("while reconciling master resources: %w", err)
+	}
+
+	results.ActualStatefulSets = actualStatefulSets
+	return results, nil
+}
+
+func maybeUpscaleMasterResources(ctx upscaleCtx, masterResources []nodespec.Resources) error {
+	// Upscale master StatefulSets using the adjusted resources and read the current StatefulSet
+	// from k8s to get the latest state.
+	for _, res := range masterResources {
+		stsName := res.StatefulSet.Name
+
+		// Read the current StatefulSet from k8s to get the latest state
+		var actualSset appsv1.StatefulSet
+		if err := ctx.k8sClient.Get(ctx.parentCtx, k8s.ExtractNamespacedName(&res.StatefulSet), &actualSset); err != nil {
+			if apierrors.IsNotFound(err) {
 				continue
 			}
+			return fmt.Errorf("while getting master StatefulSet %s: %w", stsName, err)
 		}
-		reconciled, err := es_sset.ReconcileStatefulSet(ctx.parentCtx, ctx.k8sClient, ctx.es, res.StatefulSet, ctx.expectations)
-		if err != nil {
-			return results, fmt.Errorf("reconcile StatefulSet: %w", err)
+
+		actualReplicas := sset.GetReplicas(actualSset)
+		targetReplicas := sset.GetReplicas(res.StatefulSet)
+
+		if actualReplicas < targetReplicas {
+			actualSset.Spec.Replicas = ptr.To(targetReplicas)
```

Comment:
I think we should use `UpdateReplicas(...)` instead of updating directly `Replicas`, to also update `common.k8s.elastic.co/template-hash`
```suggestion
			nodespec.UpdateReplicas(&actualSset, ptr.To[int32](targetReplicas))
```